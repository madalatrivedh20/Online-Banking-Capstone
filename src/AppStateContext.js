import { createContext, useState, useContext } from "react";

const AppStateContext = createContext();

export const AppStateProvider = ({ children }) => {
  const [isAuthenticated, setIsAuthenticated] = useState(false);
  const [issocialAuthenticated, setIssocialAuthenticated] = useState(false);
  const [user, setUser] = useState({});
  const [allUsers, setAllUsers] = useState([]);

  const value = {
    isAuthenticated,
    setIsAuthenticated,
    user,
    issocialAuthenticated,
    setIssocialAuthenticated,
    setUser,
    allUsers,
    setAllUsers
  };
  return <AppStateContext.Provider value={value}>{children}</AppStateContext.Provider>;
};

const useAppState = () => {
  const context = useContext(AppStateContext);

  if (context === undefined) {
    throw new Error("useAppState must be used within AppStateContext");
  }

  return context;
};

export default useAppState;
