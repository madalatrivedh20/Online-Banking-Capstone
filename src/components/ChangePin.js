import React, { useState, useEffect, useRef } from 'react';
import validator from 'validator';
import { toast } from 'react-toastify';
import { useNavigate } from 'react-router-dom';

import Grid from '@mui/material/Grid';
import Box from '@mui/material/Box';
import Typography from '@mui/material/Typography';
import Button from '@mui/material/Button';
import TextField from '@mui/material/TextField';

import FormHeader from './FormHeader';

import useAppContext from '../AppStateContext';
import { changeATMPIN } from '../service/api';

const ChangePin = () => {
  const navigate = useNavigate();
  const { user } = useAppContext();
  const toastId = useRef(null);

  const [data, setData] = useState({
    oldPin: "",
    newPin: "",
    confirmNewPin: "",
  });

  const [loading, setLoading] = useState(false);

  const onChangeHandler = (e) => setData({ ...data, [e.target.name]: e.target.value });

  const checkValidations = (string) =>
    [validator.isNumeric, (str) => str.length === 4].reduce((preVal, func) => preVal && func(string), true);

  const onSubmitHandler = async (e) => {
    e.preventDefault();
    if (!data.oldPin || !data.newPin || !data.confirmNewPin) return toast.error("PIN should not be empty");
    if (!checkValidations(data.oldPin) || !checkValidations(data.newPin) || !checkValidations(data.confirmNewPin))
      return toast.error("PINs should be 4 characters long");
    if (data.newPin == data.oldPin) return toast.error("New PIN should be different than old PIN");
    if (data.newPin !== data.confirmNewPin) return toast.error("New PINS dont match");

    toastId.current = toast.loading("Changing ATM PIN...");
    setLoading(true);

    const { type, render } = await changeATMPIN(user.id, data.oldPin, data.newPin);

    toast.update(toastId.current, {
      type,
      render,
      isLoading: false,
      autoClose: 5000,
      closeOnClick: true,
      pauseOnHover: true,
      draggable: true,
    });

    if (type === "success") {
      return navigate("/balance");
    } else {
      setLoading(false);
    }
  };

  return (
    // <Grid
    //   container
    //   direction="column"
    //   alignItems="center"
    //   justifyContent="center"
    //   spacing={5}>
    //   <Grid item sx={{ marginTop: '1rem', width: '100%' }}>
    //     <Typography variant="h3" align="center">
    //       Change ATM Pin
    //     </Typography>
    //   </Grid>

    //   <Grid item xs={12} alignItems="center">
    //     <TextField
    //       required
    //       label="Old PIN"
    //       variant="standard"
    //       name="oldPin"
    //       onChange={onChangeHandler}
    //       value={data.oldPin}
    //       sx={{ display: 'block' }}
    //     />
    //   </Grid>

    //   <Grid item xs={12}>
    //     <TextField
    //       required
    //       label="New Pin"
    //       variant="standard"
    //       name="newPin"
    //       value={data.newPin}
    //       onChange={onChangeHandler}
    //       sx={{ display: 'block' }}
    //     />
    //   </Grid>

    //   <Grid item xs={12}>
    //     <TextField
    //       required
    //       label="Confirm New PIN"
    //       variant="standard"
    //       name="confirmNewPin"
    //       value={data.confirmNewPin}
    //       onChange={onChangeHandler}
    //       sx={{ display: 'block' }}
    //     />
    //   </Grid>

    //   <Grid item xs={12}>
    //     <Button variant="contained"
    //       onClick={onSubmitHandler}
    //       disabled={loading}>Change</Button>
    //   </Grid>
    // </Grid >
    <div className="forms">
      <FormHeader title="Change PIN" />
      <div>
        <form onSubmit={onSubmitHandler}>
          <div className="row">
            <label>Old PIN</label>
            <input required type="text" placeholder="Enter old PIN"
              name="oldPin"
              onChange={onChangeHandler}
              value={data.oldPin} />
          </div>

          <div className="row">
            <label>New PIN</label>
            <input required type="text" placeholder="Enter new PIN"
              name="newPin"
              value={data.newPin}
              onChange={onChangeHandler} />
          </div>

          <div className="row">
            <label>Re-enter new PIN</label>
            <input required placeholder="Re-enter new PIN"
              type="text"
              name="confirmNewPin"
              value={data.confirmNewPin}
              onChange={onChangeHandler} />
          </div>

          <div id="button" className="row">
            <button>Change ATM PIN</button>

          </div>
        </form>

      </div>
    </div>
  );
};

export default ChangePin;