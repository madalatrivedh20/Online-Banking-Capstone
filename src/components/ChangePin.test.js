import { render, screen, waitForElementToBeRemoved } from '@testing-library/react';
import renderer from 'react-test-renderer';
import userEvent from '@testing-library/user-event';
import ChangePin from './ChangePin';

import {
  BrowserRouter as Router,
  Routes,
  Route
} from "react-router-dom";
import { AppStateProvider } from '../AppStateContext';

it('Snapshot Test', () => {
  const component = renderer.create(
    <AppStateProvider>
      <Router>
        <ChangePin />
      </Router>
    </AppStateProvider>,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('renders the page', async () => {
  render(
    <AppStateProvider>
      <Router>
        <ChangePin />
      </Router>
    </AppStateProvider>
  );

  expect(screen.getByRole("heading")).toHaveTextContent(/Change PIN/);
  expect(screen.getByPlaceholderText("Enter old PIN")).toBeInTheDocument();
  expect(screen.getByPlaceholderText("Enter new PIN")).toBeInTheDocument();
  expect(screen.getByPlaceholderText("Re-enter new PIN")).toBeInTheDocument();
  expect(screen.getByRole("button", { name: "Change ATM PIN" })).toBeInTheDocument();
});