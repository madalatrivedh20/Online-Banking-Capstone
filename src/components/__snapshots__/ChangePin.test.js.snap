// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Test 1`] = `
<div
  className="forms"
>
  <div>
    <h2
      id="headerTitle"
    >
      Change PIN
    </h2>
  </div>
  <div>
    <form
      onSubmit={[Function]}
    >
      <div
        className="row"
      >
        <label>
          Old PIN
        </label>
        <input
          name="oldPin"
          onChange={[Function]}
          placeholder="Enter old PIN"
          required={true}
          type="text"
          value=""
        />
      </div>
      <div
        className="row"
      >
        <label>
          New PIN
        </label>
        <input
          name="newPin"
          onChange={[Function]}
          placeholder="Enter new PIN"
          required={true}
          type="text"
          value=""
        />
      </div>
      <div
        className="row"
      >
        <label>
          Re-enter new PIN
        </label>
        <input
          name="confirmNewPin"
          onChange={[Function]}
          placeholder="Re-enter new PIN"
          required={true}
          type="text"
          value=""
        />
      </div>
      <div
        className="row"
        id="button"
      >
        <button>
          Change ATM PIN
        </button>
      </div>
    </form>
  </div>
</div>
`;
